dali/kernels/alloc.cc

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cuda_runtime.h>
+#include <cassert>
+#include "dali/kernels/alloc.h"
+#include "dali/kernels/static_switch.h"
+#include "dali/kernels/gpu_utils.h"
+
+namespace dali {
+namespace kernels {
+namespace memory {
+
+template <AllocType>
+struct Allocator;
+
+template <>
+struct Allocator<AllocType::Host> {
+  static std::function<void(void*)> GetDeallocator() {
+    return free;
+  }
+  static void *Allocate(size_t bytes) { return malloc(bytes); }
+};
+
+template <>
+struct Allocator<AllocType::Pinned> {
+  static std::function<void(void*)> GetDeallocator() {
+    int dev = 0;
+    cudaGetDevice(&dev);
+    return [dev](void *ptr) {
+      DeviceGuard guard(dev);
+      cudaFreeHost(ptr);
+    };
+  }
+
+  static void *Allocate(size_t bytes) {
+    void *ptr = nullptr;
+    cudaMallocHost(&ptr, bytes);
```

Comment:
We can add `CUDA_CALL` checks here and all the cuda calls in this file (except for cudaGetDevice)

Reply:
We cannot. CUDA_CALL has a lot of dependencies and we're in the kernels library here. However, we could throw std::bad_alloc if the call fails.